File: dmlab2d/lua/class_test.cc
// Copyright (C) 2016-2019 The DMLab2D Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "dmlab2d/lua/class.h"

#include <array>
#include <string>
#include <utility>

#include "dmlab2d/lua/bind.h"
#include "dmlab2d/lua/call.h"
#include "dmlab2d/lua/n_results_or_test_util.h"
#include "dmlab2d/lua/push_script.h"
#include "dmlab2d/lua/read.h"
#include "dmlab2d/lua/table_ref.h"
#include "dmlab2d/lua/vm.h"
#include "dmlab2d/lua/vm_test_util.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace deepmind::lab2d::lua {
namespace {

using ::deepmind::lab2d::lua::testing::IsOkAndHolds;
using ::deepmind::lab2d::lua::testing::StatusIs;
using ::testing::AllOf;
using ::testing::HasSubstr;

// Simple demo class to test and demonstrate the functionality of Class.
class Bar final : public Class<Bar> {
  friend class Class;
  static const char* ClassName() { return "system.Bar"; }

 public:
  static void Register(TableRef module, lua_State* L) {
    const Class::Reg methods[] = {};
    Class::Register(L, methods);
    module.Insert("Bar", Bar::CreateBar);
  }

  static int CreateBar(lua_State* L) {
    Class::CreateObject(L);
    return 1;
  }
};

class Foo final : public Class<Foo> {
 public:
  explicit Foo(std::string name) : name_(std::move(name)) {}

  // Foo(name): returns a new Foo object with the given name.
  static NResultsOr CreateFoo(lua_State* L) {
    std::string name;
    switch (Read(L, 1, &name).Value()) {
      case ReadResult::kFound:
        Class::CreateObject(L, std::move(name));
        return 1;
      case ReadResult::kNotFound:
        return std::string("Missing string arg1 when constructing: ") +
               ClassName();
      case ReadResult::kTypeMismatch:
      default:
        return std::string(
                   "Type missmatch arg1 is not a string when constructing: ") +
               ClassName();
    }
  }

  // Returns whatever was provided as the first argument, twice.
  NResultsOr Duplicate(lua_State* L) {
    lua_pushvalue(L, -1);
    return 2;
  }

  // Returns the name of the object.
  NResultsOr Name(lua_State* L) {
    Push(L, name_);
    return 1;
  }

  // Returns an error message.
  NResultsOr Error(lua_State* L) { return "Something went wrong!"; }

  static void Register(TableRef module, lua_State* L) {
    const Class::Reg methods[] = {
        {"duplicate", Member<&Foo::Duplicate>},  //
        {"name", Member<&Foo::Name>},            //
        {"error", Member<&Foo::Error>},          //
    };
    Class::Register(L, methods);
    module.Insert("Foo", Bind<Foo::CreateFoo>);
  }

  const std::string& name() const { return name_; }

 protected:
  friend class Class;

  static const char* ClassName() { return "system.Foo"; }

 private:
  std::string name_;
};

static int RequireFooBar(lua_State* L) {
  TableRef module = TableRef::Create(L);
  Foo::Register(module, L);
  Bar::Register(module, L);
  Push(L, module);
  return 1;
}

using ClassTest = testing::TestWithVm;

TEST_F(ClassTest, TestName) {
  TableRef module = TableRef::Create(L);
  Foo::Register(module, L);
  Push(L, "Hello");
  Foo::CreateFoo(L);
  lua_getfield(L, -1, "name");
  lua_pushvalue(L, -2);
  int error = lua_pcall(L, 1, 1, 0);
  ASSERT_EQ(error, 0) << "Reason - " << lua_tostring(L, -1);
  std::string result;
  ASSERT_TRUE(IsFound(Read(L, -1, &result)));
  EXPECT_EQ("Hello", result);
}

TEST_F(ClassTest, TestDuplicate) {
  TableRef module = TableRef::Create(L);
  Foo::Register(module, L);
  Push(L, "Hello");
  Foo::CreateFoo(L);
  lua_getfield(L, -1, "duplicate");
  lua_pushvalue(L, -2);
  Push(L, "There");
  int error = lua_pcall(L, 2, 2, 0);
  ASSERT_EQ(error, 0) << "Reason - " << lua_tostring(L, -1);
  std::string result0;
  std::string result1;
  ASSERT_TRUE(IsFound(Read(L, -1, &result0)));
  ASSERT_TRUE(IsFound(Read(L, -2, &result1)));
  EXPECT_EQ(result0, result1);
}

TEST_F(ClassTest, TestRead) {
  TableRef module = TableRef::Create(L);
  Foo::Register(module, L);
  Push(L, "Hello");
  Foo::CreateFoo(L);
  Foo* foo = Foo::ReadObject(L, -1);
  ASSERT_TRUE(foo != nullptr);
  EXPECT_EQ("Hello", foo->name());
}

constexpr char kScript2[] = R"(
local test_module = require 'test_module'
return {
    test_module.Foo('Hello'),
    test_module.Foo('Hello2'),
}
)";

TEST_F(ClassTest, TestRead2) {
  vm()->AddCModuleToSearchers("test_module", RequireFooBar);
  ASSERT_THAT(PushScript(L, kScript2, "kScript2"), IsOkAndHolds(1));
  ASSERT_THAT(Call(L, 0), IsOkAndHolds(1));
  std::array<Foo*, 2> results;
  ASSERT_TRUE(IsFound(Read(L, 1, &results)));
  EXPECT_EQ("Hello", results[0]->name());
  EXPECT_EQ("Hello2", results[1]->name());
}

constexpr char kScript[] = R"(
local test_module = require 'test_module'
return test_module.Foo('hello')
)";

TEST_F(ClassTest, TestModule) {
  vm()->AddCModuleToSearchers("test_module", RequireFooBar);
  ASSERT_THAT(PushScript(L, kScript, "kScript"), IsOkAndHolds(1));
  ASSERT_THAT(Call(L, 0), IsOkAndHolds(1));
  Foo* foo = Foo::ReadObject(L, 1);
  ASSERT_TRUE(foo != nullptr);
  EXPECT_EQ("hello", foo->name());
}

constexpr char kScriptMethodError[] = R"(
local test_module = require 'test_module'
local foo = test_module.Foo('Hello')
return foo:error()
)";

TEST_F(ClassTest, MethodErrorMessage) {
  vm()->AddCModuleToSearchers("test_module", RequireFooBar);
  ASSERT_THAT(PushScript(L, kScriptMethodError, "kScriptMethodError"),
              IsOkAndHolds(1));
  ASSERT_THAT(Call(L, 0), StatusIs(AllOf(HasSubstr("system.Foo.error"),
                                         HasSubstr("Something went wrong!"))));
}

constexpr char kScriptCallError[] = R"(
local test_module = require 'test_module'
local foo = test_module.Foo('Hello')
local bar = test_module.Bar()
-- Calling with '.'' instead of ':'
return foo.duplicate(bar)
)";

TEST_F(ClassTest, CallErrorMessage) {
  vm()->AddCModuleToSearchers("test_module", RequireFooBar);
  ASSERT_THAT(PushScript(L, kScriptCallError, "kScriptCallError"),
              IsOkAndHolds(1));
  ASSERT_THAT(Call(L, 0),
              StatusIs(AllOf(HasSubstr("system.Foo"), HasSubstr("userdata"))));
}

}  // namespace
}  // namespace deepmind::lab2d::lua

File: dmlab2d/lib/game_scripts/common/class_test.lua
--[[ Copyright (C) 2020 The DMLab2D Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
]]

local asserts = require 'testing.asserts'
local test_runner = require 'testing.test_runner'
local class = require 'common.class'
local args = require 'common.args'

local tests = {}

function tests.createClassWithoutInit()
  local MyClass = class.Class()

  function MyClass:get2()
    return 2
  end

  local myInstance = MyClass()
  asserts.EQ(myInstance:get2(), 2)
end

function tests.initCalled()
  local MyClass = class.Class()
  function MyClass:__init__()
    self._arg = 3
  end

  function MyClass:getArg()
    return self._arg
  end

  local myInstance = MyClass()
  asserts.EQ(myInstance:getArg(), 3)
end

function tests.methodsCanHaveArgs()
  local MyClass = class.Class()

  function MyClass:addFive(value)
    return 5 + value
  end

  local myInstance = MyClass()
  asserts.EQ(myInstance:addFive(0), 5)
  asserts.EQ(myInstance:addFive(1), 6)
end

function tests.initCanCallMethods()
  local MyClass = class.Class()
  function MyClass:__init__()
    self._arg = 3 + self.getArg2()
  end

  function MyClass:getArg2()
    return 2
  end

  function MyClass:addFive(value)
    return self._arg + value
  end

  local myInstance = MyClass()
  asserts.EQ(myInstance:addFive(0), 5)
  asserts.EQ(myInstance:addFive(1), 6)
end

function tests.argsForwarded()
  local MyClass = class.Class()
  function MyClass:__init__(arg0, arg1)
    self._arg0 = arg0
    self._arg1 = arg1
  end

  function MyClass:getArg0()
    return self._arg0
  end

  function MyClass:getArg1()
    return self._arg1
  end

  local myInstance = MyClass(0, 1)
  asserts.EQ(myInstance:getArg0(), 0)
  asserts.EQ(myInstance:getArg1(), 1)
end

function tests.exceptionOnInit()
  local MyClass = class.Class()
  function MyClass:__init__(arg0)
    error(arg0)
  end

  asserts.shouldFailWithMessage(function()
      local myInstance = MyClass('hello')
    end, 'hello')
end

function tests.testBaseClassMethod()
  local MyBaseClass = class.Class()

  function MyBaseClass:addFive(arg)
    return arg + 5
  end

  local MyClass = class.Class(MyBaseClass)

  function MyClass:addSix(arg)
    return self:addFive(arg) + 1
  end


  local myInstance = MyClass()
  asserts.EQ(myInstance:addFive(3), 8)
  asserts.EQ(myInstance:addSix(2), 8)
end

function tests.testBaseClassCall()
  local MyBaseClass = class.Class()

  function MyBaseClass:__init__()
    self._base = 'base'
  end

  local MyClass = class.Class(MyBaseClass)

  function MyClass:__init__()
    MyClass.Base.__init__(self)
    self._class = 'class'
  end

  local myInstance = MyClass()

  asserts.EQ(myInstance._base, 'base')
  asserts.EQ(myInstance._class, 'class')
end


function tests.testBaseClassCall2()
  local MyClass0 = class.Class()

  function MyClass0:__init__()
    self._list = {'class0'}
  end

  local MyClass1 = class.Class(MyClass0)

  function MyClass1:__init__()
    MyClass1.Base.__init__(self)
    table.insert(self._list, 'class1')
  end

  local MyClass2 = class.Class(MyClass1)

  function MyClass2:__init__()
    MyClass2.Base.__init__(self)
    table.insert(self._list, 'class2')
  end

  local myInstance0 = MyClass0()
  asserts.tablesEQ(myInstance0._list, {'class0'})

  local myInstance1 = MyClass1()
  asserts.tablesEQ(myInstance1._list, {'class0', 'class1'})

  local myInstance2 = MyClass2()
  asserts.tablesEQ(myInstance2._list, {'class0', 'class1', 'class2'})
end

function tests.testBaseClassAddMethod()
  local MyBaseClass = class.Class()

  function MyBaseClass:addSome(arg)
    return arg + 5
  end

  local MyClass = class.Class(MyBaseClass)

  function MyClass:addMore(arg)
    return self:addSome(arg) + 1
  end

  local myInstance = MyClass()

  asserts.EQ(myInstance:addMore(4), 10)

  function MyBaseClass:addSome(arg)
    return arg + 4
  end
end

function tests.testMixins()
  local MyMixin = {
      mixinFunctionInit = function(self, arg0, arg1)
        self._arg0 = arg0
        self._arg1 = arg1
      end,
      getArg0 = function(self)
        return self._arg0
      end,
      getArg1 = function(self)
        return self._arg1
      end,
  }

  local MyClass = class.Class():__include__(MyMixin)

  function MyClass:__init__(arg0, arg1)
      self:mixinFunctionInit(arg0, arg1)
  end

  local myInstance = MyClass(2, 3)

  asserts.EQ(myInstance:getArg0(), 2)
  asserts.EQ(myInstance:getArg1(), 3)
end

function tests.testTostring()
  local MyClass = class.Class():__name__('MyClass')
  asserts.EQ(tostring(MyClass), '<class \'MyClass\'>')
end

function tests.testErrorOnCtor()
  local MyClass = class.Class():__name__('MyClass')
  function MyClass:__init__(kwargs)
    kwargs = args.parse(kwargs, {{'a', args.numberType}}, 'MyClass')
  end
  asserts.shouldFail(function() MyClass{a = 'Bad'} end, 'MyClass')
end

return test_runner.run(tests)
